### ISSUE TITLE
Component source files in npm and bower package

### ISSUE THREAD

**Issue by drublic** [google/material-design-lite#793]
I suggest to add component's source files to npm and bower packages since they might be needed separately from the build.

We currently build an application where we use a Material Design based layout. We want to use MDL's components but currently don't need all of it. Apart from that we might extend certain components and want to use the code that is already available.
For this purpose it would be nice to use the original source files that were produced in this repo. For obvious reasons we want to use npm to manage the framework aso.

It would be great to have SCSS (or whatever you like) and JS source files available through npm and bower.
Is there a reason not to provide those files or do you have any plans in doing so?

Unfortunately I could not find anything about this issue other than #329 but it does not really suggests what this issue does, I think.

**Comment by addyosmani** [comment]
Thanks for the feedback. Atm, as you noted our Bower and npm packages only contain the core build artefacts (material.css, material.js and the minified + sourcemap versions of these files). 

For a more 'complete' package, we could just distribute a tagged version of master when we're generating the normal build files. This could include the full component source, build process and so on.

I think the question there would be whether this is something we would include in the default package (`npm install material-design-lite`) or is something we would make available via a separate package `npm install material-design-lite-source` or similar. 

I'll tag in @surma and @sgomes for their thoughts here, but also interested in any feedback other users have on the desire for being able to rely on the full source.

**Comment by Garbee** [comment]
If I do `npm install`, I'd expect the source files to be in there as well. We simply don't want to track them in the primary repo to keep helping people through PRs as straightforward as possible.

**Comment by surma** [comment]
I would always expect our npm/Bower sources to be consumed by a building process that strip out unused code, so I don’t see the necessity for a separate repo like `material-design-lite-source` (but maybe I am overlooking something?).

I think I would also be feasible for us to document how to build a MDL subset using our SASS files. Technically, everyone could just use [material-design-lite.scss](https://github.com/google/material-design-lite/blob/master/src/material-design-lite.scss) and comment out the components they don’t need/want. This would need some engineering love and proper documentation.

**Comment by sgomes** [comment]
I personally wouldn't expect sources in the default NPM install, particularly if there's a build process involved, so I would vote for a separate package.

Most folks would likely want to just download the compiled library, use it in their site, and keep it updated, and our current package would do just that. For everyone else, I personally would go for the `npm install material-design-lite-source` solution, because it makes it clear what the distinction between both packages is, and which one a developer should be using.

That said, this is all coming from my very limited experience with npm here, and I'm not very clear on what's the accepted standard practice in the community...

**Comment by Garbee** [comment]
hmm, the idea of `-source` being "for developers" makes good sense. My usage of npm so far is limited as well.

I am simply an advanced person who removes uneeded components, so I need the source to handle that in my build process.

I have no disagreement with going to a separate package to get sources, I just don't see the harm (aside from initial confusion of "what do I use?") in putting them in the primary package either.

**Comment by drublic** [comment]
I would consider finding the sources in the same package, which happens in other node modules from time to time. I don't know the convention for FE modules though.
I am also ok with another package. As long as devs can get the full source code via npm I am good.

**Comment by tleunen** [comment]
I don't think another module is needed for this case. You could make the source available in this repo and the bundled/minified version as well in a specific `dist` directory or something like that.
By default, the npm main file should link to source files though.

It will be easier to use only the needed files and debug if needed.

**Comment by dstaley** [comment]
I also think using a `dist` directory within the module would be the best way to go. A good example to follow would be Bootstrap. The module layout is as follows (downloaded from Bower):
```
bower_components/bootstrap/
├── Gruntfile.js
├── LICENSE
├── README.md
├── bower.json
├── dist
│   ├── css
│   │   ├── bootstrap-theme.css
│   │   ├── bootstrap-theme.css.map
│   │   ├── ...
│   ├── fonts
│   │   ├── glyphicons-halflings-regular.eot
│   │   ├── glyphicons-halflings-regular.svg
│   │   ├── ...
│   └── js
│       ├── bootstrap.js
│       ├── bootstrap.min.js
│       └── ...
├── fonts
│   ├── glyphicons-halflings-regular.eot
│   ├── glyphicons-halflings-regular.svg
│   ├── ...
├── grunt
│   ├── ...
├── js
│   ├── affix.js
│   ├── alert.js
│   ├── ...
├── less
│   ├── alerts.less
│   ├── badges.less
│   ├── ...
├── package.js
└── package.json
```

This would allow someone to easily include the built CSS file to begin, and then graduate to using the original source files without needing to install another module. It'd also allow you to use your package manager's version resolver as opposed to manually including the GitHub repo.

**Comment by derwaldgeist** [comment]
I would like to second on this, as I was also suprised that the source files had not been included in the bower package. And I also think there's no need for a separate package, since the main files can be defined in the bower.json file. Yet, I wouldn't mind if there were a separate package, like there is for bootstrap (bootstrap-sass-official instead of just bootstrap). But having the compiled files only is a deal-breaker for me. I don't want to keep the sources in a separate folder, I want all my libraries in bower, npm or jspm. There's already three different types of packaging mechanismus out there, so I am not very pleased if I have to use a direct git repo as well. Without having the sources packaged, I would have to stick to materialize.

**Comment by addyosmani** [comment]
To add more data to the conversation: if we either include sources in the default package or via a new separate one, the size of the final binary pulled down goes from under 300KB up to ~27MB (on initial looks) - this includes dist, sources, docs, templates, images etc. I'll double check this. Would the size increasing be of concern to anyone requesting source inclusion?

**Comment by tleunen** [comment]
Don't put the docs, templates and images. Just the sources and compiled
files. The size of just the sass and js files should not be that big :)

**Comment by drublic** [comment]
What @tleunen says. Otherwise I am good with some more data.

**Comment by dstaley** [comment]
If you remove `docs`, `templates`, `test`, and the `snippets` directories, the resulting size would be around 1.6MB, with 1.1MB of that size coming from the precompiled files that are already being included in the package.

**Comment by derwaldgeist** [comment]
I wouldn't need the docs, templates etc. either. The main purpose for me would be to have the Sass sources in the packages in order to create custom CSS.

**Comment by cllns** [comment]
I want to [add the Sass](https://github.com/cllns/material_design_lite-rails/issues/5) into a ruby gem ([cllns/material_design_lite-rails](https://github.com/cllns/material_design_lite-rails)). It's hard though because I want the versions to match, and right now I have to go through and find [the commit](https://github.com/google/material-design-lite/commit/0c25c12ef2641f2b129a90386d77aac8d9caf3ad) that defines the version number. 

A source tag, or switching to a `dist` folder, would be ideal, from my perspective.

I think doing the former, in a backwards compatible fashion, might be as easy as:
```
git tag source-v1.0.0 9caf3ad
```

(and add `-a`, if you want to do an annotated tag)
and pushing that tag.

I think `source-v1.0.0` is a good name, so that all the source and version tags will be grouped, since they'll be sorted alphabetically.

**Comment by addyosmani** [comment]
I think for 1.0.1 (coming week or so) we can try to add support for the core package (or a separate one) just containing the sass sources (sans docs and templates). We'll need to carefully craft this and might tag in some of you on an RC package just to make sure it contains what you need :)

**Comment by drublic** [comment]
Sounds good. We will test this. Thanks.

**Comment by cllns** [comment]
Agreed, thanks!

**Comment by zeroxme** [comment]
+1 for sources

**Comment by Ninir** [comment]
+1 :)

**Comment by hebbet** [comment]
that is an assigned issue with a milestone, so please stop adding +1.

**Comment by simonewebdesign** [comment]
Looking forward to this. Having the source files when I do `npm install` is a must for me, as I want to be able to use only what I need.

A big feature for me is also the ability to get all the defined variables and mixins, so I can customize the build directly from source instead of doing it manually from the website.

I don't mind if you want to distribute the sources in another npm package, but I'm surprised as well that the sources are not available to download. I had to setup npm to get the git repository, which is less straightforward than `npm install --save-dev material-design-lite`.

**Comment by Garbee** [comment]
We are going to roll the sources right into the main package as I recall. So no worries there unless something breaks on it being carried out.

**Comment by faviouz** [comment]
+1. Looking forward to this.

**Comment by surma** [comment]
Will be closed by #1147.

**Comment by mikemclin** [comment]
FYI, you can do this to get all source files in Bower.  The only downside is that you are now referencing an exact tagged version, and don't get the benefits of automatically updating, like you would with a `~1.0.0` or `1.0.*` version.  You would need to change the tag link manually. 

`bower install material-design-lite=https://github.com/google/material-design-lite/archive/v1.0.0.zip --save`

**Comment by addyosmani** [comment]
We're preparing to release 1.0.1, which should also include better support for installing/building the source files for MDL (excluding the docs/templates/other bits). If interested in this feature, could you help us test it out by trying:

```sh
npm install git://github.com/google/material-design-lite#release
```

This should install material-design-lite in your node_modules directory. The root of this will include our normal built dist files, but we now include a `src` directory, gulpfile.js and package.json that you can use. An `npm install` in here allows you to build your own MDL, but you can alternatively just import in the relevant files you want to use in your own Sass.

Looking for feedback on whether this is what you had in mind and if this adequately nails the use case you're using MDL for. Cheers!

**Comment by simensen** [comment]
@addyosmani Trying this package out for the first time tonight and wanted to include the source. Saw that this most recent message had been posted just minutes before. Now, less than 10 minutes later, I can say that it looks like it works great for my purposes. Thanks! (oops, i forgot to click "comment", my bad. :) )

**Comment by addyosmani** [comment]
Thanks for trying it out @simensen :)

**Comment by drublic** [comment]
Thanks for providing this solution. I've tried this build within my current project where we installed `master` before.
It works as expected and does not break any of the existing structure. Apart from that it provides the files we need to build the components we really need and extend those. The parts we've overwritten and configured (for example variables) still work as expected.

So all in all I am very happy with this move :) Thank you!
